### DISCUSSION TITLE
How to find GO-CAM models curated by UniProt for a given UniProtKB accession?

### DISCUSSION THREAD

**Discussion by dlrice** [geneontology/helpdesk#511] [Q&A]
I am trying to find the API call to fetch the GO-CAM models curated by the UniProt group for a given UniProtKB accession. I see that it's possible to filter by group on the front end website:

https://geneontology.cloud/browse/g:UniProt

But how do I modify this request:

https://api.geneontology.org/api/gp/UniProtKB:P05067/models

to only include models which are curated by the UniProt group?

**Comment by dlrice** [comment]
Hi @sierra-moxon, @kltm do you have any advice on the above question? Many thanks!

**Comment by kltm** [comment]
@dlrice Unfortunately, I think the short of this is that the GO API does not currently support this use case. While you can get models per ID, you cannot get models per group/resource.

Depending on what you're trying to do, we have a project getting under way that may be of interest to you and covers some of this use case: https://github.com/geneontology/project-management/issues/47 . We're hoping to have this exposed as part of the release process in the next few months.

Finally, the easiest sanctioned way to get this information at this point in time (the .cloud site and related APIs are deprecated) would be to download the GPAD for your resource and extract the models individual annotations are derived from from the extensions column.
If your resource is allowing the GO-CAM models to be exposed directly, they'd be at https://current.geneontology.org/annotations; otherwise, we currently expose the raw generated GPAD files at http://current.geneontology.org/products/upstream_and_raw_data/, with a resources models like noctua_fb.gpad.gz .

**Reply by dlrice** [reply]
Thanks @kltm. We're coming at this from the front-end side of things where we want to determine which GO-CAM model ID choices to offer a user within a uniprotkb entry page. Upon user selection, this GO-CAM ID will be loaded into [wc-gocam-viz](https://github.com/geneontology/wc-gocam-viz). It sounds like there is no single GO API request at the moment and rather than loading data into our backend for this I have come up with a multi-step process:

1. When a user visits https://uniprot.org/uniprotkb/P05067 fetch all possible GO-CAMs for that uniprotkb accession from https://api.geneontology.org/api/gp/UniProtKB:P05067/models eg:

```
[
  {
    "gocam": "http://model.geneontology.org/5fce9b7300002436",
    "title": "TREM2 activated by APP"
  },
  {
    "gocam": "http://model.geneontology.org/SYNGO_2279",
    "title": "SYNGO_BP_2279"
  }
]
```

2. For each model listed fetch the model information. Eg https://api.geneontology.org/api/go-cam/gomodel:5fce9b7300002436 has:

```
{
  "id": "gomodel:5fce9b7300002436",
  "individuals": [...],
  "facts": [...],
  "annotations": [
    ...
    {
      "key": "providedBy",
      "value": "https://www.uniprot.org"
    }
  ]
}
```

3. If within the annotations an item with key of `providedBy` has value `https://www.uniprot.org` consider this a uniprot curated model.

Am I interpreting `providedBy` correctly as "curated by"? Is that as about as efficient (in terms of number of requests) as I can be with the GO API at the moment?

**Comment by kltm** [comment]
@dlrice Okay, I think I understand what you're trying to do here, but maybe I'm a little unclear on the exact flow. As we have on the Alliance gene pages, wouldn't a call to the GO API to get the models by ID be sufficient? Or are you trying to further filter the models returned to a smaller subset (i.e. exclusively "assigned by" UniProt)? Looking at the models represented in `noctua_uniprotkb.gpad.gz`, I can indeed see some overlap there.

Given that, if 1) the filtering is required and 2) you only wanted to use remote APIs, then "yes", I think your flow is about the best way to accomplish something like that.

@pgaudet Given this use case, perhaps we could consider adding this to the work that @sierra-moxon will be undertaking soon with the new models? I suspect we'll be running into more stuff like this heading forward.

**Comment by dlrice** [comment]
> Or are you trying to further filter the models returned to a smaller subset (i.e. exclusively "assigned by" UniProt)?

@kltm Yes, exactly this.

> the .cloud site and related APIs are deprecated

I plan on advertising this link https://geneontology.cloud/browse/g:UniProt to users to browse more UniProt curated GO-CAMs. You mentioned earlier that this site is deprecated, is there a more up to date link I should use instead?

**Comment by kltm** [comment]
@dlrice Just checking on this smaller subset. Is the issue trying to get models for display to the curation group or for display on your pages of only UniProt-generated model, or is it a filter for model "quality" or some other property?

Also: yes, that site and domain is being deprecated and generally does not see much use. We have a project to fold this functionality into AmiGO, but that is still getting under way, so there is no new link for that functionality at this point.

**Reply by dlrice** [reply]
> Just checking on this smaller subset. Is the issue trying to get models for display to the curation group or for display on your pages of only UniProt-generated model, or is it a filter for model "quality" or some other property?

@kltm it's to display on our pages only UniProt-generated models. E.g. you can see on [our dev site](https://wwwdev.ebi.ac.uk/uniprot/front-end/dev/uniprotkb/P05067/entry) an example. If you scroll down to and open the GO-CAM Models tab it will load all of the models from https://api.geneontology.org/api/gp/UniProtKB:P05067/models and through subsequent calls to https://api.geneontology.org/api/go-cam/gomodel:... `SYNGO_BP_2279` will be removed leaving only the UniProt curated "TREM2 activated by APP" GO-CAM model for the user to select from.

> Also: yes, that site and domain is being deprecated and generally does not see much use. We have a project to fold this functionality into AmiGO, but that is still getting under way, so there is no new link for that functionality at this point.

We'll keep an eye out for your update site.
